### PR TITLE
[state_dict][4/N] Support strict flag for model.load_state_dict

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111111
* #111120
* #111110
* __->__ #111109
* #111275
* #111107
* #111106

As title

Differential Revision: [D50209723](https://our.internmc.facebook.com/intern/diff/D50209723/)

### CODE REVIEW

File: test/distributed/checkpoint/test_state_dict.py [hunk 1]
```diff
@@ -355,3 +355,18 @@ def init_model_optim():
...
         self._test_save_load(init_model_optim)
+
+    @skip_if_lt_x_gpu(1)
+    def test_strict(self) -> None:
+        model = CompositeParamModel(device=torch.device("cuda"))
+
+        model_state_dict, _ = state_dict(model)
+        key = next(iter(model_state_dict.keys()))
+        model_state_dict.pop(key)
+        with self.assertRaisesRegex(RuntimeError, "Missing key"):
```
Comment (rohan-varma) [inline test/distributed/checkpoint/test_state_dict.py:366]: might also wanna test unexpected keys, i.e. an extra key in the state dict to load?

File: test/distributed/checkpoint/test_state_dict.py [hunk 2]
```diff
@@ -355,3 +355,18 @@ def init_model_optim():
...
+            load_state_dict(model, model_state_dict=model_state_dict)
+        load_state_dict(
```
Comment (rohan-varma) [inline test/distributed/checkpoint/test_state_dict.py:368]: I believe the nn.module.load_state_dict will return `missing_keys, unexpected_keys` tuple, will this API do the same?

Comment (fegin) [reply test/distributed/checkpoint/test_state_dict.py:368]: No, it will not. We can add the support.

**Review by wz337 — APPROVED** [review]
Accepted in phabricator.

File: test/distributed/checkpoint/test_state_dict.py (hunk 2 above)
Comment (fegin) [reply test/distributed/checkpoint/test_state_dict.py:370]: I'll add the return values in a separate PR.

**Pull request closed without merging**